src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,700 @@
+      useArray = false;
+      return *this;
+    }
+    fields.emplace_back();
+    useArray = fields.size() == 1;
+    return *this;
+  }
+};
+
+// Create an adjacency list with edges from supertype to subtypes.
+std::vector<std::vector<size_t>>
+createSubtypeGraph(const std::vector<HeapType>& types) {
+  std::unordered_map<HeapType, size_t> indices;
+  for (auto type : types) {
+    indices.insert({type, indices.size()});
+  }
+
+  std::vector<std::vector<size_t>> subtypeGraph(types.size());
+  for (size_t i = 0; i < types.size(); ++i) {
+    if (auto super = types[i].getDeclaredSuperType()) {
+      if (auto it = indices.find(*super); it != indices.end()) {
+        subtypeGraph[it->second].push_back(i);
+      }
+    }
+  }
+  return subtypeGraph;
+}
+
+struct RecGroupInfo;
+
+// As we iterate through the strongly connected components, we may find
+// components that have the same shape. When we find such a collision, we merge
+// the components into a single equivalence class where we track how we have
+// disambiguated all such isomorphic components.
+struct GroupClassInfo {
+  // If the group has just a single type, record it so we can make sure the
+  // brand is not identical to it.
+  std::optional<HeapType> singletonType;
+  // If we have gone through all the permutations of this group, we need to add
+  // a brand type to continue differentiating different groups in the class.
+  std::optional<BrandTypeIterator> brand;
+  // An adjacency list giving edges from supertypes to subtypes within the
+  // group, using indices into the (non-materialized) canonical shape of this
+  // group, offset by 1 iff there is a brand type. Used to ensure that we only
+  // find emit permutations that respect the constraint that supertypes must be
+  // ordered before subtypes.
+  std::vector<std::vector<size_t>> subtypeGraph;
+  // A generator of valid permutations of the components in this class.
+  TopologicalOrders orders;
+
+  // Initialize `subtypeGraph` and `orders` based on the canonical ordering
+  // encoded by the group and permutation in `info`.
+  static std::vector<std::vector<size_t>> initSubtypeGraph(RecGroupInfo& info);
+  GroupClassInfo(RecGroupInfo& info);
+
+  void advance() {
+    ++orders;
+    if (orders == orders.end()) {
+      advanceBrand();
+    }
+  }
+
+  void advanceBrand() {
+    if (brand) {
+      ++*brand;
+    } else {
+      brand.emplace();
+      // Make room in the subtype graph for the brand type, which goes at the
+      // beginning of the canonical order.
+      subtypeGraph.insert(subtypeGraph.begin(), {{}});
+      // Adjust indices.
+      for (size_t i = 1; i < subtypeGraph.size(); ++i) {
+        for (auto& edge : subtypeGraph[i]) {
+          ++edge;
+        }
+      }
+    }
+    // Make sure the brand is not the same as the real type.
+    if (singletonType &&
+        RecGroupShape({**brand}) == RecGroupShape({*singletonType})) {
+      ++*brand;
+    }
+    // The brand type must be distinct from
```

Comment:
comment fragment?

Reply:
Yep, looks like an unfinished predecessor to "// Make sure the brand is not the same as the real type" above. Will delete.